packages/core/src/codewhispererChat/storages/chatHistory.ts

```diff
@@ -118,22 +115,81 @@ export class ChatHistoryManager {
             this.clearRecentHistory()
         }
 
-        if (this.history.length <= MaxConversationHistoryLength) {
-            return
+        // Check if we need to trim based on character count
+        const totalCharacters = this.calculateHistoryCharacterCount()
+        if (totalCharacters > MaxConversationHistoryCharacters) {
+            this.logger.debug(
+                `History size (${totalCharacters} chars) exceeds limit of ${MaxConversationHistoryCharacters} chars`
+            )
+            // Keep removing messages from the beginning until we're under the limit
+            do {
+                // Find the next valid user message to start from
+                const indexToTrim = this.findIndexToTrim()
+                if (indexToTrim !== undefined && indexToTrim > 0) {
+                    this.logger.debug(
+                        `Removing the first ${indexToTrim} elements in the history due to character count limit`
+                    )
+                    this.history.splice(0, indexToTrim)
+                } else {
+                    // If we can't find a valid starting point, reset it
+                    this.logger.debug('Could not find a valid point to trim, reset history to reduce character count')
+                    this.history = []
+                }
+            } while (
+                this.calculateHistoryCharacterCount() > MaxConversationHistoryCharacters &&
+                this.history.length > 2
+            )
         }
+    }
 
-        const indexToTrim = this.findIndexToTrim()
-        if (indexToTrim !== undefined) {
-            this.logger.debug(`Removing the first ${indexToTrim} elements in the history`)
-            this.history.splice(0, indexToTrim)
-        } else {
-            this.logger.debug('No valid starting user message found in the history, clearing')
-            this.history = []
+    private calculateHistoryCharacterCount(): number {
```

Comment:
can we make use of the logic here? https://github.com/aws/aws-toolkit-vscode/blob/master/packages/core/src/codewhispererChat/controllers/chat/chatRequest/converter.ts#L34

Reply:
I think the usecase are too different and the history truncation/handling already exists in chatHistory.ts